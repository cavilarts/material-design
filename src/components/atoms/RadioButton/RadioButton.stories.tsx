import { Meta, Story } from '@storybook/react/types-6-0';

import RadioButton, { radioButtonProps } from './index';

const radioSettings = {
  component: RadioButton,
  title: 'Atoms/RadioButton',
};

export default radioSettings as Meta;

export const radioButton: Story<radioButtonProps> = () => {
  return <RadioButton options={['Option 1', 'Option 2']} onClick={() => {}} />;
};
