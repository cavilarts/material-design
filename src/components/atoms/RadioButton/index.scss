.radio {
  display: flex;
  outline: none;
  margin: 5px;

  &__wrapper {
    box-sizing: border-box;
    display: flex;
    width: 20px;
    height: 20px;
    position: relative;
  }

  &__outerCircle {
    box-sizing: border-box;
    display: block;
    height: 20px;
    left: 0;
    position: absolute;
    top: 0;
    transition: border-color ease 280ms;
    width: 20px;
    border-width: 2px;
    border-style: solid;
    border-radius: 50%;
    border-color: rgba(0, 0, 0, 0.54);

    &--selected {
      border-color: #ff4081;
    }
  }

  &__innerCircle {
    background-color: #ff4081;
    border-radius: 50%;
    box-sizing: border-box;
    height: 20px;
    width: 20px;
    position: absolute;
    top: 0;
    transition: transform ease 280ms, background-color ease 280ms;
    transform: scale(0.001);

    &--selected {
      transform: scale(0.5);
    }
  }

  &__input {
    visibility: hidden;
  }

  &__labelContent {
    user-select: auto;
    display: flex;
    order: 0;
    line-height: inherit;
    padding-left: 8px;
  }
}
