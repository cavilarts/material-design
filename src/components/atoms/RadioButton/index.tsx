import { useCallback, useState } from 'react';
import classNames from 'classnames';

import './index.scss';
export interface radioButtonProps {
  options: string[];
  onClick: (index: number) => void;
}

const RadioButton: React.FC<radioButtonProps> = ({ options, onClick }) => {
  const [selectedIndex, setSelectedIndex] = useState(0);

  const clickCallback = useCallback(
    index => {
      setSelectedIndex(index);
      onClick(index);
    },
    [onClick],
  );

  return (
    <div className='radio__listWrapper'>
      {options.map((x, index) => (
        <label htmlFor={`radio-${index}`} key={index} className='radio'>
          <span className='radio__wrapper'>
            <span
              className={classNames('radio__outerCircle', { 'radio__outerCircle--selected': selectedIndex === index })}
            ></span>
            <span
              className={classNames('radio__innerCircle', { 'radio__innerCircle--selected': selectedIndex === index })}
            ></span>
            <input
              className='radio__input'
              type='radio'
              id={`radio-${index}`}
              name={x}
              value={x}
              onChange={() => {
                clickCallback(index);
              }}
              checked={selectedIndex === index}
            />
          </span>
          <span className='radio__labelContent'>{x}</span>
        </label>
      ))}
    </div>
  );
};

export default RadioButton;
