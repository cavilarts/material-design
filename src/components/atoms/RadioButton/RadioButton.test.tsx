import { shallow, configure } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

import RadioButton from './index';

configure({ adapter: new Adapter() });

describe('Radio Button', () => {
  it('Should render a radio button element', () => {
    const radioButton = shallow(<RadioButton options={['Radio Option']} onClick={() => {}} />);

    expect(radioButton).toMatchSnapshot();
  });

  it('should select the clicked option', () => {
    const clickCallback = jest.fn();
    const radioButton = shallow(<RadioButton options={['Radio Option', 'Another Option']} onClick={clickCallback} />);

    radioButton.find('input').at(1).simulate('click', {});
    expect(clickCallback).toHaveBeenCalledWith(1);
  });
});
