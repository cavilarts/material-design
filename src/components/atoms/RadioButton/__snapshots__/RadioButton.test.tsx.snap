// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio Button Should render a radio button element 1`] = `
<div
  className="radio_listWrapper"
>
  <label
    className="radio"
    htmlFor="radio-0"
    key="0"
  >
    <span>
      <span />
      <span />
      <input
        checked={true}
        id="radio-0"
        name="Radio Option"
        onClick={[Function]}
        type="radio"
        value="Radio Option"
      />
    </span>
    <span>
      Radio Option
    </span>
  </label>
</div>
`;
